src/python/pants/goal/run_tracker.py

```diff
@@ -390,3 +413,40 @@ def shutdown_worker_pool(self):
     N.B. This exists only for internal use and to afford for fork()-safe operation in pantsd.
     """
     SubprocPool.shutdown(self._aborted)
+
+  def report_target_info(self, scope, target, keys, val):
+    """Add target information to run_info under target_data.
+
+    Will Recursively construct a nested dict with the keys provided.
```

Comment:
Am I right in observing that in practice, keys is always a singleton? If so, do we need the complexity of the recursion?